### PR TITLE
Issue2079

### PR DESCRIPTION
**Pull request by singhshraddha**
**issue** https://github.ibm.com/wiotp/Maximo-Asset-Monitor/issues/2079
**further discussion** https://ibm.ent.box.com/notes/772880075609

**Testcase**
Tested with 5 different signals (as shown below)

image

image

### CODE REVIEW

File: iotfunctions/data_quality.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+logger = logging.getLogger(__name__)
+
+PACKAGE_URL = 'git+https://github.com/ibm-watson-iot/functions.git@'
+_IS_PREINSTALLED = True
+
+
+class DataQualityChecks(BaseComplexAggregator):
+    """
+    Data Quality module will help asses the quality of incoming sensor data, using the provided metrics.
```
Comment (pkohlmann) [inline iotfunctions/data_quality.py:26]: I'm not a native speaker but I think a fourth 's' at the tail of 'asses' prevents misunderstandings :-)

Comment (pkohlmann) [reply iotfunctions/data_quality.py:26]: fyi, 'asses' occurs more often in the description... :-)

**Review by pkohlmann — CHANGES_REQUESTED** [review]

File: iotfunctions/data_quality.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        return stationary_type[adf_stationary, kpss_stationary]
+
+    @staticmethod
+    def stuck_at_zero(series):
+        """
+        A time series signal stuck at zero contains no information
+
+        :returns bool
+        """
+        return bool(np.count_nonzero(series.to_numpy()) < 1)
```
Comment (sedgewickmm18) [inline iotfunctions/data_quality.py:166]: How about using np.all_close to account for values that are almost zero (up to an error of a 10^-8 or so)

File: iotfunctions/data_quality.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            m += 1
+            xm = np.array([L[i: i + m] for i in range(N - m + 1)])
+
+            A = np.sum([np.sum(np.abs(xmi - xm).max(axis=1) <= r) - 1 for xmi in xm])
+
+            # Return SampEn
+            return -np.log(A / B)
+
+        return sampen(series.to_list(), m=2, r=0.2 * series.std())
+
```
Comment (sedgewickmm18) [inline iotfunctions/data_quality.py:119]: Cool, I only knew about approximate entropy.

File: requirements.txt [hunk 4]
```diff
@@ -18,3 +18,4 @@ tabulate==0.8.5
 pyarrow==3.0.0
 stumpy==1.5.1
 torch>=1.3.1
+statsmodels>=0.12
```
Comment (sedgewickmm18) [inline requirements.txt:21]: we already include statsmodels==0.11.1 (in line with WML) and I think all tests you use are included into that version

**Review by sedgewickmm18 — CHANGES_REQUESTED** [review]
Please check whether you can go with statsmodels 0.11.1

Thumbs up for coding around DQA. 
We don't need their data imputation stuff right now and that's where they put most of their efforts in.

File: requirements.txt (hunk 4 above)
Comment (singhshraddha) [reply requirements.txt]: done

**Review by pkohlmann — APPROVED** [review]

**Pull request merged**